src/test/java/com/billingapi/controller/InvoiceControllerTest.java

```diff
@@ -0,0 +1,586 @@
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.BDDMockito.given;
+import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+
+
+@WebMvcTest(InvoiceController.class)
+public class InvoiceControllerTest {
+    @Autowired
+    private MockMvc mockMvc;
+
+    @MockBean
+    private InvoiceService invoiceService;
+
+    @Autowired
+    private ObjectMapper objectMapper;
+    private Object BigDecimal;
+
+
+
+ /*  Test to Create invoice if Method is not allowed then it throws exception  HttpRequestMethodNotSupported
+  */
+    @Test
+    public void should_throw_exception_whenCreateInvoice_ifHttpRequestMethodNotSupported() throws Exception {
+
+        InvoiceRequest newInvoice = InvoiceRequest.builder()
+                .itemName("jeans")
+                .price(new BigDecimal("98000.0"))
+                .tax(10.00)
+                .vat(15.00)
+                .discount(10.00)
+                .isItemOnSale(true)
+                .discountOnSale(10.00)
+                .quantity(999999990)
+                .total(new BigDecimal("98000.0"))
+                .description("good")
+
+                .build();
+
+        given(invoiceService.save(any(Invoice.class)))
+                .willAnswer((invocation) -> invocation.getArgument(0));
+
+        // when - action or behaviour that we are going test
+        ResultActions response = mockMvc.perform(put("/invoice")
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(objectMapper.writeValueAsString(newInvoice)));
+
+        // then - verify the result or output using assert statements
+
+        response.andDo(print()).
+                andExpect(status().isMethodNotAllowed());
+
+    }
+
+
+    /*  Test to Create invoice if Method arguments is not valid   then it throws exception  MethodArgumentTypeMismatch
+     */
+    @Test
+    public void should_throw_exception_whenCreateInvoice_ifMethodArgumentTypeMismatch() throws Exception {
+
+        // given - precondition or setup
+                  int invoiceId = 1;
+                  Invoice newInvoice = Invoice.builder()
+                .itemName("jeans")
+                .price(new BigDecimal("98000.0"))
+                .tax(10.00)
+                .vat(15.00)
+                .discount(10.00)
+                .isItemOnSale(true)
+                .discountOnSale(10.00)
+                .userId(1)
+                .isDeleted(false)
+                .timestamp(LocalDateTime.now())
+                .description("good")
+                .quantity(2)
+                .total(new BigDecimal("98000.0"))
+                .build();
+
+        given(invoiceService.findInvoiceById(invoiceId)).willReturn(newInvoice);
+
+        // when -  action or the behaviour that we are going test
+        ResultActions response = mockMvc.perform(get("/invoice/aa", invoiceId));
```

Comment:
declare a variable for url e.g. INVALID_INVOICE_ID_URL = "/invoice/aa" . To make the code more readble